pytensor/tensor/rewriting/elemwise.py

```diff
@@ -529,333 +539,486 @@ def local_upcast_elemwise_constant_inputs(fgraph, node):
-                            pass
-
-                    inputs.append(i)
-                    s_inputs.append(s)
-                s_g.append(s)
-
-        if not fused:
-            return False
-
-        if new_nb_input != len(inputs) or len(s_inputs) != len(inputs):
-            # TODO FIXME: This shouldn't be a generic `Exception`
-            raise Exception(
-                "Something has gone wrong with the elemwise fusion rewrite; skipping."
-            )
-
-        s_new_out = node.op.scalar_op(*s_g, return_list=True)
-        try:
-            s_new_out[0].owner.op.c_code(
-                s_new_out[0].owner,
-                "test_presence_of_c_code",
-                ["x" for x in s_g],
-                ["z" for x in s_new_out],
-                {"fail": "%(fail)s"},
-            )
-        except (NotImplementedError, MethodNotDefined):
-            name = str(s_new_out[0].owner.op)
-            warn(
-                "Rewrite warning: "
-                f"The Op {name} does not provide a C implementation."
-                " As well as being potentially slow, this also disables "
-                "loop fusion."
-            )
-            return False
-
-        # create the composite op.
-        composite_op = aes.Composite(s_inputs, s_new_out)
-
-        # create the new node.
-        # Do not call make_node to have test_value
-        new_node = maker(node, composite_op)(*inputs).owner
-
-        assert len(new_node.outputs) == 1
-        assert node.outputs[0].type.dtype == new_node.outputs[0].type.dtype
+                    new_scalar_input = aes.get_scalar_type(
+                        inp.type.dtype
+                    ).make_variable()
+                    node_scalar_inputs.append(new_scalar_input)
+                    middle_scalar_inputs.append(new_scalar_input)
+                    middle_inputs.append(inp)
+
+            new_scalar_node = node.op.scalar_op.make_node(*node_scalar_inputs)
+            middle_scalar_inputs.append(new_scalar_node.outputs[0])
+            middle_inputs.append(node.outputs[0])
+
+        scalar_outputs = [
+            middle_scalar_inputs[middle_inputs.index(out)] for out in fg.outputs
+        ]
+        return scalar_inputs, scalar_outputs
 
-        if len(new_node.inputs) > max_nb_input:
-            warn(
-                "Loop fusion failed because the resulting node "
-                "would exceed the kernel argument limit."
-            )
-            return False
-
-        # we fuse as many that we can at the same time to make debug mode faster
-        # debug mode will be faster as it won't test all intermediate step.
-        while True:
-            ret = local_fuse(fgraph, new_node)
-            if ret is not False and ret is not None:
-                assert len(ret) == len(new_node.outputs)
-                assert len(ret) == 1
-                new_node = ret[0].owner
-            else:
-                break
+    def apply(self, fgraph):
```

Comment:
this should be broken down into functions - at least at the `while` and `for` levels

Reply:
I broke it down, so that the main loop looks like:

```python
            fuseable_clients, unfuseable_clients = initialize_stateful_mappings(fg=fg)
            visited_nodes: Set[Apply] = set()
            while True:
                starting_nodes = fg.apply_nodes.copy()
                try:
                    subgraph_inputs, subgraph_outputs = find_fuseable_subgraph(
                        fg=fg,
                        visited_nodes=visited_nodes,
                        fuseable_clients=fuseable_clients,
                        unfuseable_clients=unfuseable_clients,
                    )
                except ValueError:
                    return
                else:
                    update_stateful_mappings_after_fg_update(
                        fg=fg,
                        visited_nodes=visited_nodes,
                        fuseable_clients=fuseable_clients,
                        unfuseable_clients=unfuseable_clients,
                        starting_nodes=starting_nodes,
                    )
```
WDYT?

Reply:
The main loop looks much better, but the local functions make it still really hard to read, as it is ~450 lines in total.
Would it be feasible to extract these local functions and pass a few more things as args/kwargs if needed?

Your call though

Reply:
I know it's hard to read but I am quite averse to adding functions that will only be called just once. It's certainly more readable but at some point the runtime overhead of another function in the stack doesn't make sense. 

I tried an iteration with more functions and it wasn't mich better to have a function with 6 other internal functions inside.

I tried to add comments to explain what's going on. That and collapsing scopes in an IDE will hopefully be enough to make it readable.

The code would be much simpler with:
1) A stateless algo would be much more concise, but it was prohibitively slow
2) Some specialized graph objects that have internal methods to evaluate the constraints that matter here would make the code much more readable, but I don't want to go down that path just for the sake of this rewrite
3) An alternative algorithm, if someone has ideas